Engine/Component/ComponentBillboard.cpp

```diff
@@ -0,0 +1,235 @@
+{
+	alignment_type = _alignment_type;
+
+	if (alignment_type == SPRITESHEET)
+		is_spritesheet = true;
+	else
+		is_spritesheet = false;
+}
+
+void ComponentBillboard::Render(const float3& position)
+{
+	GLuint shader_program = App->program->GetShaderProgramId("Billboard");
+	glUseProgram(shader_program);
+
+	int n;
+	glGetProgramStageiv(shader_program, GL_VERTEX_SHADER, GL_ACTIVE_SUBROUTINE_UNIFORM_LOCATIONS, &n);
+	unsigned* subroutines_indices = new unsigned[n];
+
+	//Subroutine functions
+	GLuint viewpoint_subroutine = glGetSubroutineIndex(shader_program, GL_VERTEX_SHADER, "view_point_alignment");
+	GLuint crossed_subroutine = glGetSubroutineIndex(shader_program, GL_VERTEX_SHADER, "crossed_alignment");
+	GLuint axial_subroutine = glGetSubroutineIndex(shader_program, GL_VERTEX_SHADER, "axial_alignment");
+
+	//Subroutine uniform
+	int selector = glGetSubroutineUniformLocation(shader_program, GL_VERTEX_SHADER, "alignment_selector");
+
+	switch (alignment_type) {
+	case VIEW_POINT:
+		glUniformSubroutinesuiv(GL_VERTEX_SHADER, n, &viewpoint_subroutine);
+		break;
+	case CROSSED:
+		glUniformSubroutinesuiv(GL_VERTEX_SHADER, n, &crossed_subroutine);
+		break;
+
+	case AXIAL:
+		glUniformSubroutinesuiv(GL_VERTEX_SHADER, n, &axial_subroutine);
+		break;
+
+	case SPRITESHEET:
+		if(oriented_to_camera)
+			glUniformSubroutinesuiv(GL_VERTEX_SHADER, n, &viewpoint_subroutine);
+
+		else
+			glUniformSubroutinesuiv(GL_VERTEX_SHADER, n, &crossed_subroutine);
+
+		glUniform1i(glGetUniformLocation(shader_program, "billboard.XTiles"), x_tiles);
+		glUniform1i(glGetUniformLocation(shader_program, "billboard.YTiles"), y_tiles);
+		glUniform1f(glGetUniformLocation(shader_program, "X"), current_sprite_x);
+		glUniform1f(glGetUniformLocation(shader_program, "Y"), current_sprite_y);
+
+		glUniform1f(glGetUniformLocation(shader_program, "billboard.speed"), sheet_speed);
+		SwitchFrame();
+		break;
+
+	default:
+		// viewpoint by default, the most consistent one
+		glUniformSubroutinesuiv(GL_VERTEX_SHADER, n, &viewpoint_subroutine);
+		break;
+	}
+
+	glActiveTexture(GL_TEXTURE0);
+	glBindTexture(GL_TEXTURE_2D, billboard_texture->opengl_texture);
+	glUniform1i(glGetUniformLocation(shader_program, "billboard.texture"), 0);
+	glUniform1f(glGetUniformLocation(shader_program, "billboard.width"), width);
+	glUniform1f(glGetUniformLocation(shader_program, "billboard.height"), height);
+	glUniform1f(glGetUniformLocation(shader_program, "billboard.isSpritesheet"), is_spritesheet);
+	glUniform4fv(glGetUniformLocation(shader_program, "billboard.color"),1, (float*)color);
+	glUniform3fv(glGetUniformLocation(shader_program, "billboard.center_pos"), 1, position.ptr());
+
+
+	glBindVertexArray(vao);
+	glDrawElements(GL_TRIANGLES, 6, GL_UNSIGNED_INT, 0);
+	glBindVertexArray(0);
+
+	glUseProgram(0);
+}
+
+Component* ComponentBillboard::Clone(bool original_prefab) const
+{
+	ComponentBillboard * created_component;
```

Comment:
code style  ComponentBillboard* instead of ComponentBillboard *

Reply:
done